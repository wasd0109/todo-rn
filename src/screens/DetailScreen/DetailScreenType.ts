import { RouteProp } from "@react-navigation/native";
import { StackNavigationProp } from "@react-navigation/stack";
import { RootStackParamList } from "../../types/ReactNavigationType";

type DetailScreenNavigationProp = StackNavigationProp<
  RootStackParamList,
  "Detail"
>;
type DetailScreenRouteProp = RouteProp<RootStackParamList, "Detail">;

export type DetailScreenProps = {
  navigation: DetailScreenNavigationProp;
  route: DetailScreenRouteProp;
};
