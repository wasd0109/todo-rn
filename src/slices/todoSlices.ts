import { createSlice, PayloadAction } from "@reduxjs/toolkit";

export interface Todo {
  title: string;
  date: number;
  id: string;
}

export interface TodoState {
  todoList: Todo[];
}

const initialState: TodoState = {
  todoList: [],
};

export const todoSlice = createSlice({
  name: "todo",
  initialState,
  reducers: {
    addTodo: (state, action: PayloadAction<Todo>) => {
      state.todoList.push(action.payload);
    },
  },
});

export const { addTodo } = todoSlice.actions;

export default todoSlice.reducer;
