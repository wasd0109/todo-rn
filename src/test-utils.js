import React from "react";
import "react-native-paper";
import { render } from "@testing-library/react-native";
import { Provider as PaperProvider } from "react-native-paper";
import { NavigationContainer } from "@react-navigation/native";
import { createStackNavigator } from "@react-navigation/stack";

const AllTheProviders = ({ children }) => {
  return <PaperProvider>{children}</PaperProvider>;
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react-native";

// override render method
export { customRender as render };

const Stack = createStackNavigator();
const MockedNavigator = ({ screenName, component, params = {} }) => {
  return (
    <NavigationContainer>
      <Stack.Navigator>
        <Stack.Screen
          name={screenName}
          component={component}
          initialParams={params}
        />
      </Stack.Navigator>
    </NavigationContainer>
  );
};

export default MockedNavigator;
