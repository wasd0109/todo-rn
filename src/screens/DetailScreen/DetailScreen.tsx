import React, { useState } from "react";
import { StyleSheet, Text, View } from "react-native";
import { Button, TextInput } from "react-native-paper";
import { useSelector } from "react-redux";
import { RootState } from "../../store";
import { format } from "date-fns";
import Spacer from "../../components/Spacer/Spacer";

import { DetailScreenProps } from "./DetailScreenType";

const DetailScreen = ({ navigation, route }: DetailScreenProps) => {
  const todoList = useSelector((state: RootState) => state.todo.todoList);
  const todo = todoList.find((todo) => todo.id === route.params.id);
  const [newTitle, setNewTitle] = useState(todo?.title);
  if (!todo)
    return (
      <View style={styles.errorContainerStyle}>
        <Text style={styles.errorMessageStyle}>Something wrong happened</Text>
      </View>
    );
  return (
    <View style={styles.containerStyle}>
      <Spacer>
        <TextInput
          value={newTitle}
          testID="title-input"
          onChangeText={setNewTitle}
          label="Title"
          placeholder={todo.title}
        />
      </Spacer>
      <Spacer>
        <Text>Created at: {format(todo.date, "mm/dd/yy kk:mm")}</Text>
      </Spacer>
      <Spacer>
        <Button mode="contained" style={styles.buttonStyle}>
          Edit
        </Button>
      </Spacer>
      <Spacer>
        <Button mode="outlined" color="red" style={styles.deleteButtonStyle}>
          Delete
        </Button>
      </Spacer>
    </View>
  );
};

export default DetailScreen;

const styles = StyleSheet.create({
  errorMessageStyle: { fontSize: 30 },
  errorContainerStyle: {
    justifyContent: "center",
    flex: 1,
    alignSelf: "center",
  },
  containerStyle: {
    flex: 1,
    justifyContent: "center",
    marginHorizontal: 15,
    alignContent: "center",
  },
  buttonStyle: {
    width: 200,
    alignSelf: "center",
  },
  deleteButtonStyle: {
    alignSelf: "center",
    borderColor: "red",
    width: 200,
  },
});
