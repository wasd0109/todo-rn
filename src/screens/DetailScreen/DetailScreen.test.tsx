import React from "react";
import { render } from "../../test-utils";
import DetailScreen from "./DetailScreen";
import { fireEvent, RenderAPI } from "@testing-library/react-native";
import { configureStore, createSlice } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { v4 as uuidv4 } from "uuid";
import { format } from "date-fns";
import { PayloadAction } from "@reduxjs/toolkit";
import { Todo } from "../../slices/todoSlices";

const TEST_TITLE = "Hello";
const TODO_ID = uuidv4();

const initialState = {
  todoList: [{ title: "Test", date: 123456, id: TODO_ID }],
};

const todoSlice = createSlice({
  name: "todo",
  initialState,
  reducers: {
    editTodo: (state, action: PayloadAction<Todo>) => {
      const index = state.todoList.findIndex(
        (todo) => todo.id === action.payload.id
      );
      state.todoList.splice(index, 1, action.payload);
    },
    deleteTodo: (state, action: PayloadAction<string>) => {
      const index = state.todoList.findIndex(
        (todo) => todo.id === action.payload
      );
      state.todoList.splice(index, 1);
    },
  },
});
const store = configureStore({ reducer: { todo: todoSlice.reducer } });

const props = {
  navigation: { goBack: jest.fn() },
  route: {
    params: { id: TODO_ID },
  },
} as any;

let component: RenderAPI;

beforeEach(() => {
  component = render(
    <Provider store={store}>
      <DetailScreen {...props} />
    </Provider>
  );
});

describe("DetailScreen render correctly", () => {
  test("DetailScreen render", () => {
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("DetailScreen show date created correctly ", () => {
    const { getAllByText } = component;
    const date = initialState.todoList[0].date;
    expect(
      getAllByText("Created at: " + format(date, "mm/dd/yy kk:mm")).length
    ).toEqual(1);
  });
});

describe("DetailScreen function properly", () => {
  test("DetailScreen allow Todo title to be edited", () => {
    const { getByTestId, getAllByDisplayValue } = component;
    const titleInputRef = getByTestId("title-input");
    fireEvent.changeText(titleInputRef, TEST_TITLE);
    expect(getAllByDisplayValue(TEST_TITLE).length).toEqual(1);
  });

  test("DetailScreen edit button edit the Todo title", () => {
    const { getByTestId, getByText } = component;
    const titleInputRef = getByTestId("title-input");
    const editBtnRef = getByText("Edit");
    fireEvent.changeText(titleInputRef, TEST_TITLE);
    fireEvent.press(editBtnRef);
    const todoArray = store
      .getState()
      .todo.todoList.filter((todo) => todo.title === TEST_TITLE);
    expect(todoArray.length).toEqual(1);
  });

  test("DetailScreen delete button delete the Todo title", () => {
    const { getByTestId, getByText } = component;
    const titleInputRef = getByTestId("title-input");
    const delBtnRef = getByText("Delete");
    fireEvent.changeText(titleInputRef, TEST_TITLE);
    fireEvent.press(delBtnRef);
    const todoArray = store
      .getState()
      .todo.todoList.filter((todo) => todo.title === TEST_TITLE);
    expect(todoArray.length).toEqual(0);
  });
});
