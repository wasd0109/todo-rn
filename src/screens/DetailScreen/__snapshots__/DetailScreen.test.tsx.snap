// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailScreen render correctly DetailScreen render 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignContent": "center",
        "flex": 1,
        "justifyContent": "center",
        "marginHorizontal": 15,
      }
    }
  >
    <View
      style={
        Object {
          "marginVertical": 10,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "rgb(231, 231, 231)",
              "borderTopLeftRadius": 4,
              "borderTopRightRadius": 4,
            },
            Object {},
          ]
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "rgba(0, 0, 0, 0.26)",
              "bottom": 0,
              "height": 2,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "transform": Array [
                Object {
                  "scaleY": 0.5,
                },
              ],
            }
          }
        />
        <View
          style={
            Array [
              Object {
                "paddingBottom": 0,
                "paddingTop": 0,
              },
              Object {
                "minHeight": 64,
              },
            ]
          }
        >
          <View
            pointerEvents="none"
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "opacity": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "transform": Array [
                  Object {
                    "translateX": 3,
                  },
                ],
              }
            }
          >
            <Text
              numberOfLines={1}
              onLayout={[Function]}
              style={
                Object {
                  "color": "#6200ee",
                  "fontFamily": "System",
                  "fontSize": 16,
                  "fontWeight": undefined,
                  "left": 0,
                  "opacity": 0,
                  "paddingLeft": 12,
                  "paddingRight": 12,
                  "position": "absolute",
                  "textAlign": "left",
                  "top": 34,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                    Object {
                      "translateY": -16,
                    },
                    Object {
                      "scale": 0.75,
                    },
                  ],
                  "writingDirection": "ltr",
                }
              }
            >
              Title
            </Text>
            <Text
              numberOfLines={1}
              style={
                Object {
                  "color": "rgba(0, 0, 0, 0.54)",
                  "fontFamily": "System",
                  "fontSize": 16,
                  "fontWeight": undefined,
                  "left": 0,
                  "opacity": 0,
                  "paddingLeft": 12,
                  "paddingRight": 12,
                  "position": "absolute",
                  "textAlign": "left",
                  "top": 34,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                    Object {
                      "translateY": -16,
                    },
                    Object {
                      "scale": 0.75,
                    },
                  ],
                  "writingDirection": "ltr",
                }
              }
            >
              Title
            </Text>
          </View>
          <TextInput
            allowFontScaling={true}
            editable={true}
            multiline={false}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder=""
            placeholderTextColor="rgba(0, 0, 0, 0.54)"
            rejectResponderTermination={true}
            selectionColor="#6200ee"
            style={
              Array [
                Object {
                  "flexGrow": 1,
                  "margin": 0,
                  "zIndex": 1,
                },
                Object {
                  "paddingLeft": 12,
                  "paddingRight": 12,
                },
                Object {
                  "height": 64,
                },
                Object {
                  "paddingBottom": 4,
                  "paddingTop": 24,
                },
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontSize": 16,
                  "fontWeight": undefined,
                  "textAlign": "left",
                  "textAlignVertical": "center",
                },
                false,
                Array [
                  Object {},
                ],
              ]
            }
            testID="title-input"
            underlineColorAndroid="transparent"
            value="Test"
          />
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "marginVertical": 10,
        }
      }
    >
      <Text>
        Created at: 
        02/01/70 09:02
      </Text>
    </View>
    <View
      style={
        Object {
          "marginVertical": 10,
        }
      }
    >
      <View
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "#6200ee",
            "borderColor": "transparent",
            "borderRadius": 4,
            "borderStyle": "solid",
            "borderWidth": 0,
            "elevation": 2,
            "minWidth": 64,
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 0.75,
              "width": 0,
            },
            "shadowOpacity": 0.24,
            "shadowRadius": 1.5,
            "width": 200,
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "disabled": undefined,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 4,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "letterSpacing": 1,
                      "marginHorizontal": 16,
                      "marginVertical": 9,
                      "textAlign": "center",
                    },
                    undefined,
                    Object {
                      "textTransform": "uppercase",
                    },
                    Object {
                      "color": "#ffffff",
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    Object {
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    undefined,
                  ],
                ]
              }
            >
              Edit
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "marginVertical": 10,
        }
      }
    >
      <View
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "transparent",
            "borderColor": "red",
            "borderRadius": 4,
            "borderStyle": "solid",
            "borderWidth": 0.5,
            "elevation": 0,
            "minWidth": 64,
            "width": 200,
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "disabled": undefined,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 4,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "letterSpacing": 1,
                      "marginHorizontal": 16,
                      "marginVertical": 9,
                      "textAlign": "center",
                    },
                    undefined,
                    Object {
                      "textTransform": "uppercase",
                    },
                    Object {
                      "color": "red",
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    Object {
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    undefined,
                  ],
                ]
              }
            >
              Delete
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
